src/libproc_macro/bridge/client.rs

```diff
@@ -0,0 +1,450 @@
+    pub(super) run: extern "C" fn(Bridge, F, Buffer<u8>) -> Buffer<u8>,
+    pub(super) f: F,
+}
+
+// FIXME(#53451) public to work around `Cannot create local mono-item` ICE,
+// affecting not only the function itself, but also the `BRIDGE` `thread_local!`.
+pub extern "C" fn __run_expand1(
+    bridge: Bridge,
+    f: fn(::TokenStream) -> ::TokenStream,
+    mut b: Buffer<u8>,
+) -> Buffer<u8> {
+    panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        bridge.enter(|| {
+            let input = {
+                let b = &mut &b[..];
+                TokenStream::decode(b, &mut ())
+            };
+
+            let output = f(::TokenStream(input)).0;
+
+            // HACK(eddyb) Separate success (vs panics) encoding to avoid
+            // bringing handles outside the `bridge.enter(|| ...)` scope.
+            b.clear();
+            Ok::<_, ()>(output).encode(&mut b, &mut ());
+        })
+    })).map_err(PanicMessage::from)
+    .unwrap_or_else(|e| {
+        b.clear();
+        Err::<(), _>(e).encode(&mut b, &mut ());
+    });
+    b
+}
+
+impl Client<fn(::TokenStream) -> ::TokenStream> {
+    pub const fn expand1(f: fn(::TokenStream) -> ::TokenStream) -> Self {
+        Client {
+            get_handle_counters: HandleCounters::get,
+            run: __run_expand1,
+            f,
+        }
+    }
+}
+
+// FIXME(#53451) public to work around `Cannot create local mono-item` ICE,
+// affecting not only the function itself, but also the `BRIDGE` `thread_local!`.
+pub extern "C" fn __run_expand2(
+    bridge: Bridge,
+    f: fn(::TokenStream, ::TokenStream) -> ::TokenStream,
+    mut b: Buffer<u8>,
+) -> Buffer<u8> {
+    panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        bridge.enter(|| {
+            let (input, input2) = {
+                let b = &mut &b[..];
+                (
+                    TokenStream::decode(b, &mut ()),
+                    TokenStream::decode(b, &mut ()),
+                )
+            };
+
+            let output = f(::TokenStream(input), ::TokenStream(input2)).0;
+
+            // HACK(eddyb) Separate success (vs panics) encoding to avoid
+            // bringing handles outside the `bridge.enter(|| ...)` scope.
+            b.clear();
+            Ok::<_, ()>(output).encode(&mut b, &mut ());
+        })
+    })).map_err(PanicMessage::from)
+    .unwrap_or_else(|e| {
+        b.clear();
+        Err::<(), _>(e).encode(&mut b, &mut ());
+    });
+    b
+}
+
+impl Client<fn(::TokenStream, ::TokenStream) -> ::TokenStream> {
+    pub const fn expand2(f: fn(::TokenStream, ::TokenStream) -> ::TokenStream) -> Self {
+        Client {
+            get_handle_counters: HandleCounters::get,
+            run: __run_expand2,
+            f,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub enum ProcMacro {
+    CustomDerive {
+        trait_name: &'static str,
+        attributes: &'static [&'static str],
+        client: Client<fn(::TokenStream) -> ::TokenStream>,
```

Comment:
The `fn` parameter was making me a bit worried, so I just wanted to make sure to clarify this part. The `::TokenStream` here is an *abi unsafe* type, right? Despite that, though, `Client` here is totally fine because `fn(...) -> ...` is ABI safe (just a pointer). The actual function here is only ever invoked on the client, never on the server. On the server it's purely opaque in that it's just passed around.

Is that all right?